BIP 32 binary encoding functions are extracted from base58

dr-orlovsky opened the pull request:
Binary (non-base58) encoding for public and private keys are required for storing PSBT `PSBT_GLOBAL_XPUB` according to BIP 174. This PR extracts binary encoding from inside `Display` and `FromStr` functions and put this as a standalone methods.

This PR is dependent on #469; to see only changes specific to the current PR pls check commit https://github.com/rust-bitcoin/rust-bitcoin/commit/94c2ddaeffd4277aed50bc88e0684e4d311faeea

stevenroose commented:
Could this perhaps be rebased just on master? I think this can be merged faster than the ypub stuff as it's way less opinionated on.

dr-orlovsky commented:
@stevenroose done

stevenroose reviewed (changes requested):
I don't understand one of the From cases and I think one error can be `expect`'ed.

dr-orlovsky commented on src/util/bip32.rs:
```diff
@@ -489,6 +510,57 @@ impl ExtendedPrivKey {
+            network: network,
+            depth: data[4],
+            parent_fingerprint: Fingerprint::from(&data[5..9]),
+            child_number: child_number,
```
Just didn't wanted to touch original code (apart from moving it to a standalone function) to simplify the review. Done with special commit.

And I have nothing against `.into()` when it is obvious from the context into which type the conversion happens

dr-orlovsky commented on src/util/bip32.rs:
```diff
@@ -489,6 +510,57 @@ impl ExtendedPrivKey {
+            private_key: PrivateKey {
+                compressed: true,
+                network: network,
+                key: secp256k1::SecretKey::from_slice(
```
Agree

elichai replied:
It also verifies that the key isn't zero and isn't overflowing

dr-orlovsky replied:
Fixed & force-pushed

stevenroose replied:
Hmm, does it check it's not zero? I thought that was only for messages?

stevenroose replied:
Hmm, ack the Ecdsa error return.

dr-orlovsky commented:
Rebased onto fixed master (requires #512 to be merged first) and addressed @stevenroose comments

dr-orlovsky commented:
Found another incorrect error conversion which fixed with https://github.com/rust-bitcoin/rust-bitcoin/pull/470/commits/97e7d0da8a04a2c79319c755735f24a812915958

apoelstra commented on src/util/bip32.rs:
```diff
@@ -489,6 +510,57 @@ impl ExtendedPrivKey {
+        } else if data[0..4] == [0x04u8, 0x35, 0x83, 0x94] {
+            Network::Testnet
+        } else {
+            return Err(base58::Error::InvalidVersion((&data[0..4]).to_vec()).into());
```
Did you mean to use the new `UnknownVersion` error variant?

apoelstra commented on src/util/bip32.rs:
```diff
@@ -531,16 +531,14 @@ impl ExtendedPrivKey {
             depth: data[4],
             parent_fingerprint: Fingerprint::from(&data[5..9]),
-            child_number: child_number,
+            child_number:  endian::slice_to_u32_be(&data[9..13]).into(),
```
extra space added here

apoelstra commented on src/util/bip32.rs:
```diff
@@ -434,7 +434,10 @@ impl From<base58::Error> for Error {
+            other => base58::Error::Other(other.to_string())
+        }
     }
 }
```
This whole impl is weird and should be deleted.

dr-orlovsky commented:
@apoelstra your comments are addressed. Rebased onto master with the merged fix and kept two commits in the history: one for moving logic from old functions to a new once w/o code modification inside those functions; and the other on top with all improvements and addressing review suggestions.

apoelstra commented on src/util/bip32.rs:
```diff
@@ -587,106 +702,41 @@ impl ExtendedPubKey {
-        ret[45] = 0;
-        ret[46..78].copy_from_slice(&self.private_key[..]);
-        fmt.write_str(&base58::check_encode_slice(&ret[..]))
+        fmt.write_str(&base58::check_encode_slice(&self.encode()[..]))
```
Can you use https://docs.rs/bitcoin/0.25.2/bitcoin/util/base58/fn.check_encode_slice_to_fmt.html to avoid the allocation

stevenroose commented on src/util/bip32.rs:
```diff
@@ -572,7 +644,50 @@ impl ExtendedPubKey {
+            child_number: child_number,
+            chain_code: ChainCode::from(&data[13..45]),
+            public_key: PublicKey::from_slice(
+                &data[45..78])?,
```
nit: No reason for a newline here, but going to ack anyway

dr-orlovsky replied:
Don't know how it happened, probably occasionally hit Enter... Fixed

stevenroose commented on src/util/bip32.rs:
```diff
@@ -572,7 +644,50 @@ impl ExtendedPubKey {
+        }
+
+        let cn_int: u32 = endian::slice_to_u32_be(&data[9..13]);
+        let child_number: ChildNumber = ChildNumber::from(cn_int);
```
nit: Type doesn't need to be specified twice.

dr-orlovsky replied:
True, remnants on the older code

stevenroose commented on src/util/bip32.rs:
```diff
@@ -572,7 +644,50 @@ impl ExtendedPubKey {
+            },
+            depth: data[4],
+            parent_fingerprint: Fingerprint::from(&data[5..9]),
+            child_number: child_number,
```
In fact this one can be inlined just like the above one to `endian::slice_to_u32_be(&data[9..13]).into()`

stevenroose reviewed (changes requested):
Looks good, could you perhaps do that one extra inline? Feel free to ignore the two nits otherwise.

dr-orlovsky commented:
Updated according to @stevenroose review; all the changes are in https://github.com/rust-bitcoin/rust-bitcoin/pull/470/commits/fa4ecb4a6a8cab36bef92a0e792cdddd5b395c1a

stevenroose reviewed (approved):
LGTM
